### PR TITLE
Cookoff - Mini-Rewrite

### LINKED ISSUE #8683: Ace advanced vehicle damage / cookoff creates smoke pillar in wrong location
Mods (complete and add to the following information):

    Arma 3: 2.06.148221
    CBA: 3.15.6
    ACE3: 3.14.0

Description:
Ace 3 cookoff creates an additional smoke pillars slightly in front of a vehicle when it is destroyed via cookoff
Smoke rises from the vehicle hull and from the gun (intended) but there is an aditional smoke source floating in front of the vehicle

Steps to reproduce:
place tank of ifv
destroy it to cause cookoff
a smoke pillar will be created ca. 1m in front of the vehicle

Expected behavior:
Smoke should rise from the vehicle hull and gun (intended)
Smoke should not rise from in front of the vehicle

Where did the issue occur?
When an armored vehicle cooks off, both vanilla and modded (RHS) vehicles are affected

Screenshots:
examples of the bug appearing when various vehicles cook of
as you can see there is an additional smoke pillar in front of the vehicle that should not be present
APC - smoke rises from gun and hull additional bugged smoke source in fromt of vehicle

Tank - smoke rises from gun and hull and end of barrel - bugged smoke source half way down the barrel

IFV -  smoke rises from gun and hull additional bugged smoke source in fromt of vehicle

### LINKED ISSUE #8736: Vehicles will only cook off once
**Mods:**
- **Arma 3:** `2.06.148470 stable`
- **CBA:** `3.15.6.211004 stable`
- **ACE3:** `3.14.1 stable`

**Description:**
If a vehicle cooks off, then repair and rearm it, it will not cook off again.

If Enable ammunition cook off is active it can happen 2 things:
The vehicle cooks off, you can then repair and rearm it, if its hit again the vehicle won't cook off, but ther will be an ammunition cook off and the vehicle ammo and the ammo in the inventory will still be there. Some times the ammo cooks off at the same time as the vehicle.

or

There is an ammunition cook off, the vehicle ammo and the ammo in the inventory will still be there and the vehicle will not cook off again.

**Steps to reproduce:**
1. place a tank in the editor
2. shoot it until it cooks off, to get only an ammo cook off try to not hit any module in the vehicle, but its not reliable.
3. repair and rearm it
4.  shoot at it again

**Expected behavior:**
The vehicle should cook off as normal.

**Where did the issue occur?**
- Dedicated and Editor (Singleplayer)

**Additional context:**
I have seen it on vanilla and RHS vehicles. On dedicated it was tested with several mods and in a liberation. On editor the vehicles were repaired and rearmed with Zeus.

Edit:
Tested again on dedicated just with CBA and ACE, still happening. I don't know if i'm explaining the difference between vehicle and ammo cook off right.

### LINKED ISSUE #9743: Cook off bug - vehicle never explodes, and cook off flame occurs multiple times.
**Mods:**
- **Arma 3:** `2.14 profiling
- **CBA:** `3.16.1
- **ACE3:** `3.16.2 stable

**Description:**
Cook off doesn't seem to be working properly. It occurs continuously with short breaks, and sometimes it also looks bugged like some kind of star pattern. The vehicle will never explode even though the "destroy vehicles after cook off" is selected. Advanced vehicle damage is on as well.

**Steps to reproduce:**
Launch Arma, with ace
Spawn vehicle using zeus, attempt to destroy it via various means.

**Expected behavior:**
I expected the vehicle to cook off and then explode.

**Where did the issue occur?**
- Dedicated / Self-Hosted Multiplayer / Singleplayer / Editor (Singleplayer) / Editor (Multiplayer) - the issue occurs in every of these cases.

**Log Files:**
https://pastebin.com/7QJe48ie

**Additional context:**
None.

**Screenshots:**
image
image

### PR DESCRIPTION
**Pull request by johnb432**
**When merged this pull request will:**
- Fixes #8736.
- Fixes #8683 by placing the smoke pillar on the engine selection.
- To start vehicle or ammo box cook-offs, an event has to be run on the server.
- Same goes for ammo cook-offs.
- Made all effects JIP compatible ~~(exception are the jet and ring effects, as they only last 20 seconds max)~~.
- Removed tiny explosion effects that were present in the smoke when an ammo box is cooking off, as it looks excessive when machine gun ammo is detonating. If projectiles are created during cook-off, they will ensure that such effects occur, but in a more suited manner to the cook-off happening.
- Ammo boxes can now cook-off from incendiary rounds (incendiary definition comes from `vehicle_damage`).
- Attached the cook-off sound to the vehicle, so that it moves with the vehicle.
- Cook-off sound is deleted when the vehicle is deleted.
- CBA settings have been reworked:
  - `GVAR(enable)` has been removed.
    - This setting was too ambiguous in my opinion. 3 separate settings exist (`GVAR(enableFire)`, `GVAR(enableAmmoCookoff)` & `GVAR(enableAmmobox)`) that allow to control most aspects that `GVAR(enable)` does, making it redundant.
    - With the current system, if players dismount the vehicle that is going to cook off fast enough, the vehicle won't cook off. Checking for players within a vehicle is therefore pointless in my eyes.
    - This change means that setting `QGVAR(enable)` on objects will no longer affect ammo detonation, but only fire effects. `QGVAR(enableAmmoCookoff)` already exists and can be set in the editor, although that's currently not listed in the docs - that has been rectified.
  - `GVAR(enableFire)` is now the sole setting that handles vehicle cook-off fire. This setting does not handle ammo detonation (handled by `GVAR(enableAmmoCookoff)`).
  - Added a duration coefficient setting `GVAR(cookoffDuration)` for cook-off (fixes #9743).
  - `GVAR(probabilityCoef)` has been made useful again. It's only used in the `advanced vehicle damage` component.
  - Moved `GVAR(removeAmmoDuringCookoff)` from `vehicle_damage` into `cookoff`.
    Changed it so that it removes all magazines upon ammunition cook-off.
- Code cleanup.

SP and MP testing has been conducted, but could use some more.

### IMPORTANT

- If the contribution affects [the documentation](https://github.com/acemod/ACE3/tree/master/docs), please include your changes in this pull request so the documentation will appear on the [website](https://ace3.acemod.org/).
- [Development Guidelines](https://ace3.acemod.org/wiki/development/) are read, understood and applied.
- Title of this PR uses our standard template `Component - Add|Fix|Improve|Change|Make|Remove {changes}`.

### CODE REVIEW

File: docs/wiki/framework/events-framework.md
Comment (johnb432) [inline docs/wiki/framework/events-framework.md:116]: I'm unsure about what to do: I'm not sure if these events should be public or not.

File: addons/cookoff/functions/fnc_cookOffBoxLocal.sqf
Comment (rautamiekka) [inline addons/cookoff/functions/fnc_cookOffBoxLocal.sqf:80]: **Idea:** Not every mag, and individual cartridge (or the like) in a mag, will be gone at the same time, so it would be better if the contained mags would be iterated and given a random time to get destroyed, which would allow to save at least some ammo if you're quick enough, which makes it a lot more forgiving and realistic.

**Comment by Drofseh** [comment]
Something that has always frustrated me about vehicle cookoff is who the ammo goes shooting out of the vehicle, sometimes landing 100m+ away from it. A round's propellant burning doesn't actually give it velocity unless it's contained by a barrel, it will just pop out of its casing, and ammo in storage shouldn't be exiting the vehicle at all.

File: addons/cookoff/functions/fnc_cookOffEffect.sqf
Comment (lambdatiger) [inline addons/cookoff/functions/fnc_cookOffEffect.sqf:87]: This was a problem before your refactor, but the force addition here is tied to framerate instead of game time. This results in undesirable behavior when time acceleration isn't equal to one. It's a particular problem for factors less than 0.1, including a factor of 0, which doesn't cause `isGamePaused` to return true. 
There is probably a better approach, but a solution I've been doing locally is this:
```sqf 
if (_jet && !isGamePaused && { local _obj }) then {
    private _nextForceTime = _obj getVariable [QGVAR(nextForceTime), 0];
    if (_nextFlameTime > CBA_missionTime) then {
        private _force = [0, 0, _factor * -(0.5 min random 1.5) * (0.3 min random 1)] vectorMultiply getMass _obj;
        _obj addForce [_force, vectorUpVisual _obj];
        _obj setVariable [QGVAR(nextForceTime), _nextForceTime + 0.015]; // 15ms/circa 67 Hz limit
    };
};
```

File: addons/cookoff/functions/fnc_detonateAmmunition.sqf
Comment (Apricot-ale) [inline addons/cookoff/functions/fnc_detonateAmmunition.sqf:145]: Is it possible to change the sound to be played back to read from cfgSounds?
I would like to make it easier to create a MOD that changes cookoff's sounds.

File: addons/cookoff/functions/fnc_cookOffEffect.sqf
Comment (johnb432) [reply addons/cookoff/functions/fnc_cookOffEffect.sqf:87]: > undesirable behavior
Can you elaborate on that?

File: addons/cookoff/functions/fnc_detonateAmmunition.sqf
Comment (johnb432) [reply addons/cookoff/functions/fnc_detonateAmmunition.sqf:145]: I want to get #5327 merged, which changes the sounds.
It would be pointless doing doing that sort of thing here, as most config entries would change, but I think we can add something to #5327 to make it configurable (would go through `CfgSFX` though).

File: addons/cookoff/functions/fnc_cookOffBoxLocal.sqf
Comment (johnb432) [reply addons/cookoff/functions/fnc_cookOffBoxLocal.sqf:80]: Yes, but as it is, `FUNC(detonateAmmunition)` handles ammunition detonation entirely, disregarding the actual magazines in the vehicle/box.
It would be more realistic, but it would also be more performance intensive, as well as require some non-trivial redesign, meaning it's not guaranteed to happen.

Comment (rautamiekka) [reply addons/cookoff/functions/fnc_cookOffBoxLocal.sqf:80]: A shame, but fair enough.

File: addons/cookoff/functions/fnc_cookOffEffect.sqf
Comment (lambdatiger) [reply addons/cookoff/functions/fnc_cookOffEffect.sqf:87]: Sorry for not being clearer. When time acceleration is small or 0 (such as when the splendid camera is first opened, or in some slowed events), vehicles cooking off (if `_jet` is true) still recieve a force down every frame. The result of this per-frame force can be a number of things but often vehicles will either have a sudden collision with the ground and send themselves flying or clip through all together, i.e.,

https://github.com/acemod/ACE3/assets/11843718/e11d02c8-7104-47ba-9186-0c1a1ac794e2
Note time acceleration in top right.

**Comment by LinkIsGrim** [comment]
This or #5327 first for review?

**Comment by johnb432** [comment]
> This or #5327 first for review?

#5327 should probably be first, although that means I need to include the changes I mentioned above.

**Comment by johnb432** [comment]
> > This or #5327 first for review?
> 
> #5327 should probably be first, although that means I need to include the changes I mentioned above.

Done, #5327 is ready to review.

File: addons/cookoff/functions/fnc_cookOffBoxLocal.sqf
Comment (johnb432) [reply addons/cookoff/functions/fnc_cookOffBoxLocal.sqf:80]: It's not going to happen, until BI fixes https://community.bistudio.com/wiki/magazineTurretAmmo and https://community.bistudio.com/wiki/setMagazineTurretAmmo. I rewrote what was necessary and added an extra function (see spoiler), but imo it's too intensive performance wise.

<details>
<summary>Function</summary>

```sqf
#include "..\script_component.hpp"
/*
 * Author: johnb43
 * Removes a magazine from the vehicle that has a specific ammo count.
 *
 * Arguments:
 * 0: Vehicle <OBJECT>
 * 1: Magazine <STRING>
 * 2: Turret path <ARRAY>
 * 3: Ammo count <NUMBER>
 *
 * Return Value:
 * Magazine removed <BOOL>
 *
 * Example:
 * [vehicle player, "12Rnd_120mm_HEAT_MP_T_Red", 12, [0]] call ace_common_fnc_removeSpecificMagazineTurret
 *
 * Public: No
 */

params [["_vehicle", objNull, [objNull]], ["_magazineType", "", [""]], ["_turretPath", [], [[]]], ["_ammoCount", 0, [0]]];

if !(_vehicle turretLocal _turretPath) exitWith {};

// Check how much ammo a weapon has
private _magazinesAmmo = [];
private _found = false;

{
    _x params ["_xMag", "_xTurret", "_xAmmo"];

    if ((_xMag == _magazineType) && {_xTurret isEqualTo _turretPath}) then {
        _magazinesAmmo pushBack _xAmmo;
    };
} forEach (magazinesAllTurrets _vehicle);

private _index = _magazinesAmmo find _ammoCount;

// If not found, exit
if (_index == -1) exitWith {false};

private _weapon = "";

// Find weapon that has mag (assumption: 1 magazine can only fit in 1 weapon per turret)
{
    if (_magazineType in (compatibleMagazines _x)) exitWith {
        _weapon = _x;
    };
} forEach (_vehicle weaponsTurret _turretPath);

if (_weapon == "") exitWith {false};

// Get current weapon
(weaponState [_vehicle, _turretPath]) params ["_currentWeapon", "", "", "_currentMagazine", "_currentAmmo"];

_vehicle removeMagazinesTurret [_magazineType, _turretPath];
_magazinesAmmo deleteAt _index;

// If the current magazine is not the specified magazine, there is nothing to worry about, so delete as such
if (_currentMagazine != _magazineType) exitWith {
    // Add old magazines back
    {
        _vehicle addMagazineTurret [_magazineType, _turretPath, _x];
    } forEach _magazinesAmmo;

    true
};

// Remove weapon
_vehicle removeWeaponTurret [_weapon, _turretPath];

// If the magazine that's to be removed is the current magazine, take extra steps
if (!isNil {_magazinesAmmo deleteAt (_magazinesAmmo find _currentAmmo)}) then {
    // Add old magazine back, then add weapon back, so that mag is instantly loaded (otherwise engine takes highest ammo magazine)
    _vehicle addMagazineTurret [_magazineType, _turretPath, _currentAmmo];
    _vehicle addWeaponTurret [_weapon, _turretPath];

    // Add rest of old magazines back
    {
        _vehicle addMagazineTurret [_magazineType, _turretPath, _x];
    } forEach _magazinesAmmo;
} else {
    // Add rest of old magazines back
    {
        _vehicle addMagazineTurret [_magazineType, _turretPath, _x];
    } forEach _magazinesAmmo;

    _vehicle addWeaponTurret [_weapon, _turretPath];
};

// Select old weapon again
_vehicle selectWeaponTurret [_currentWeapon, _turretPath];

true

```
</details>

**Comment by johnb432** [comment]
This is fully ready to review.

File: addons/cookoff/functions/fnc_handleDamageBox.sqf
Comment (LinkIsGrim) [inline addons/cookoff/functions/fnc_handleDamageBox.sqf:37]: Is this correct? Why do we care about `_hitIndex` above then?

Comment (johnb432) [reply addons/cookoff/functions/fnc_handleDamageBox.sqf:37]: I pretty much left this code alone - I'll take a more in depth look at it though, as, admittedly, it's not very intuitive.

Comment (johnb432) [reply addons/cookoff/functions/fnc_handleDamageBox.sqf]: Latest commit should address this, although I need to retest it.

Comment (johnb432) [reply addons/cookoff/functions/fnc_handleDamageBox.sqf]: I have retested and it's all good. Latest commit also includes some minor tweaks.

File: docs/wiki/framework/events-framework.md
Comment (LinkIsGrim) [reply docs/wiki/framework/events-framework.md]: Fine as they are IMO

File: addons/cookoff/functions/fnc_cookOffEffect.sqf
Comment (lambdatiger) [reply addons/cookoff/functions/fnc_cookOffEffect.sqf]: Thanks!

**Review by LinkIsGrim — CHANGES_REQUESTED** [review]

File: addons/cookoff/stringtable.xml
Comment (johnb432) [reply addons/cookoff/stringtable.xml:20]: Are you sure about this?

My changes as is will already have some backwards compatibility 'breaking' effects, specifically: `GVAR(enable)` previously affected a mix of fire effects and ammo detonation. However, given the editor settings, this to me did not make sense, so I changed so that `GVAR(enable)` affects fire effects only.

If you still want a name change (which tbh I'm all for), I'd recommend `GVAR(enableFire)` instead, as that's an already existing setting (although one could use that as a very solid counter argument).

A problem will remain though:
https://github.com/acemod/ACE3/blob/master/docs/wiki/framework/cookoff-framework.md#1-disabling--enabling-cook-off-for-individual-vehicles
`GVAR(enable)` set on objects is API, so we shouldn't (but could if really wanted) change that.

Comment (LinkIsGrim) [reply addons/cookoff/stringtable.xml:20]: Best to keep backwards compat then, but setting name should be more specific IMO.

Comment (johnb432) [reply addons/cookoff/stringtable.xml]: Updated.

Comment (johnb432) [reply addons/cookoff/stringtable.xml]: [cc06b37](https://github.com/acemod/ACE3/pull/9758/commits/cc06b37f47adee85c080f55c0c2740a1da1a3e9c) removes `GVAR(enable)` in favor of `GVAR(enableFire)`, which should make it even more clear.

File: addons/grenades/functions/fnc_incendiary.sqf
Comment (LinkIsGrim) [inline addons/grenades/functions/fnc_incendiary.sqf:237]: Can't we skip this check since we're just raising an event here?

Comment (johnb432) [reply addons/grenades/functions/fnc_incendiary.sqf:237]: I'm not sure how it handles, but I've seen it in places in ACE. I thought it would be better practice to call the event only if the component was loaded.

Comment (LinkIsGrim) [reply addons/grenades/functions/fnc_incendiary.sqf:237]: We'd just raise an eventhandler with no listener (well, 3rd party might listen I guess, is this API?)

Comment (johnb432) [reply addons/grenades/functions/fnc_incendiary.sqf:237]: It is API in the docs.

Comment (LinkIsGrim) [reply addons/grenades/functions/fnc_incendiary.sqf:237]: Then keep the check

**Comment by johnb432** [comment]
I'll hold off from merging until #9061 is merged.

File: addons/cookoff/functions/fnc_cookOffEffect.sqf
Comment (johnb432) [inline addons/cookoff/functions/fnc_cookOffEffect.sqf:223]:
```suggestion
        (_tiWheels + _intensity * 0.004) / 1.002, // wheels/tracks are further away from burning parts
```

**Comment by johnb432** [comment]
> @johnb432 merge conflicts

I need to still resolve some stuff, so hold off from merging, I'll handle it.

**Comment by johnb432** [comment]
I'm done with the changes. Those changes are:
- redid internal events, so that API events wouldn't change
- fixed some bugs that came with merging 9061

**Pull request merged**
